### PR TITLE
Websocket en-/decoding revisited

### PR DESCRIPTION
**Pull request by icing**
Reimplementation of websocket frame handling in send/recv.

- state is fully kept at connection, since curl_ws_send() and curl_ws_rec() have lifetime beyond usual transfers
- no more limit on frame sizes
- uses own `bufq` to have less pointer acrobatics

### CODE REVIEW

**Comment by bagder** [comment]
A [failing CI job](https://ci.appveyor.com/project/curlorg/curl/builds/46819566/job/exqr6afokko22ut5):
~~~
test 2304...[WebSockets curl_ws_recv() with closed connection]
sh: line 1: 27438 Aborted                 (core dumped) ./libtest/lib2304.exe ws://127.0.0.1:56553/2304 > log/stdout2304 2> log/stderr2304
lib2304.exe returned 134, when expecting 0
 exit FAILED

**Comment by icing** [comment]
> A [failing CI job](https://ci.appveyor.com/project/curlorg/curl/builds/46819566/job/exqr6afokko22ut5):
> 
> ```
> test 2304...[WebSockets curl_ws_recv() with closed connection]
> sh: line 1: 27438 Aborted                 (core dumped) ./libtest/lib2304.exe ws://127.0.0.1:56553/2304 > log/stdout2304 2> log/stderr2304
> lib2304.exe returned 134, when expecting 0
>  exit FAILED
> ```

assertion failure. hmm...

**Review by bagder — APPROVED** [review]

File: lib/easy.c [hunk 1]
```diff
@@ -1223,6 +1223,24 @@ CURLcode curl_easy_recv(struct Curl_easy *data, void *buffer, size_t buflen,
...
+  if(result)
+    return result;
+
+  if(!data->conn)
+    /* on first invoke, the transfer has been detached from the connection and
+       needs to be reattached */
+    Curl_attach_connection(data, c);
+
+  return CURLE_OK;
+}
```
Comment (bagder) [inline lib/easy.c:1242]: I'd suggest making this `#ifdef USE_WEBSOCKETS` since it is the only user of this function.

Comment (icing) [reply lib/easy.c:1243]: Added and pushed.

**Comment by bagder** [comment]
On Appveyor:
~~~
C:\projects\curl\lib\ws.c(313,30): error C2220: the following warning is treated as an error [C:\projects\curl\lib\libcurl.vcxproj]
C:\projects\curl\lib\ws.c(313,30): warning C4132: 'tmp': const object should be initialized [C:\projects\curl\lib\libcurl.vcxproj]

**Comment by icing** [comment]
> On [Appveyor](https://ci.appveyor.com/project/curlorg/curl/builds/46868846/job/3ysljcjgvh5ebwrh):
> 
> ```
> C:\projects\curl\lib\ws.c(313,30): error C2220: the following warning is treated as an error [C:\projects\curl\lib\libcurl.vcxproj]
> C:\projects\curl\lib\ws.c(313,30): warning C4132: 'tmp': const object should be initialized [C:\projects\curl\lib\libcurl.vcxproj]
> ```

Pushed.

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**